### PR TITLE
Deploy from existing package

### PR DESCRIPTION
**Pull request by wbreza**
Adds `--from-package` param to `azd deploy` to support deploying a service from an existing package.

### CODE REVIEW

**Comment by DaveLHX** [comment]
I'm not sure this is the appropriate place to ask this but this looks like a feature I am looking for since I want to use the cli in our different deployment stages where the app is already compiled and packaged.  
Does this version also allow to run the provision and deploy commands with a configured service that has no project folder?
Because if I have the following in the current version I cant even run the "azd infra create" command because the project gives an error.

services:  
   web:
     project: ./Dist 
     dist: ./Dist
     language: dotnet
     host: appservice

**Comment by wbreza** [comment]
> I'm not sure this is the appropriate place to ask this but this looks like a feature I am looking for since I want to use the cli in our different deployment stages where the app is already compiled and packaged. Does this version also allow to run the provision and deploy commands with a configured service that has no project folder? Because if I have the following in the current version I cant even run the "azd infra create" command because the project gives an error.
> 
> services: web: project: ./Dist dist: ./Dist language: dotnet host: appservice

Can you provide more context into a service without a project folder? How do you typically produce the package that you want to deploy?

**Comment by DaveLHX** [comment]
> > I'm not sure this is the appropriate place to ask this but this looks like a feature I am looking for since I want to use the cli in our different deployment stages where the app is already compiled and packaged. Does this version also allow to run the provision and deploy commands with a configured service that has no project folder? Because if I have the following in the current version I cant even run the "azd infra create" command because the project gives an error.
> > services: web: project: ./Dist dist: ./Dist language: dotnet host: appservice
> 
> @DaveLHX Can you provide more context into a service without a project folder? How do you typically produce the package that you want to deploy?

Basically we have a stage that builds and packages the app and stores the package as an artifact.  Then for each deployment stages we deploy the infra with terraform and the code from the package.  So like the 3 stages to the right would ideally not care about the code.  Juste deploy the infra and whatever packages we have configured in the yaml. 
I guess a question it raises is what happens in these stages if I actually point to the code folder would it try to build stuff?
image

**Comment by wbreza** [comment]
> > > I'm not sure this is the appropriate place to ask this but this looks like a feature I am looking for since I want to use the cli in our different deployment stages where the app is already compiled and packaged. Does this version also allow to run the provision and deploy commands with a configured service that has no project folder? Because if I have the following in the current version I cant even run the "azd infra create" command because the project gives an error.
> > > services: web: project: ./Dist dist: ./Dist language: dotnet host: appservice

@DaveLHX - Yes, with this feature you can create azd service packages using `azd package <service>`, then later on deploy them with `azd deploy <service> --from-package <package_path>`.  This would only be available to deploy services one-by-one at the moment.

File: cli/azd/cmd/deploy.go
Comment (weikanglim) [inline cli/azd/cmd/deploy.go:70]: @savannahostrowski Thoughts on this flag? `from-package` or `from-packaged`?

Comment (savannahostrowski) [reply cli/azd/cmd/deploy.go:70]: I probably prefer `from-package`. That reads a bit nicer to me.

I guess I could see someone maaaaybe being like "what package"?, but then again you have to intentionally use this flag so that's probably me being a bit paranoid.

File: cli/azd/cmd/deploy.go
Comment (weikanglim) [reply cli/azd/cmd/deploy.go:72]: @savannahostrowski @ellismg @wbreza  I realized now that this means "zip file" or "container image tag".

I wonder this means we end up having to break this contract if we changed the "packaging" format from `zip` files into `directories`. Should we maybe consider introducing these two features at the same time? Or do we need to make `--from-package` a "beta feature" as well?

File: cli/azd/cmd/deploy.go
Comment (vhvb1989) [inline cli/azd/cmd/deploy.go:241]: It feels like `azd` should have some `validation` before using the input path/tag and invoking deploy.  it seems like a security vulnerability.
Should azd take a package that was created by azd only?  

What happen, for example, if I call `azd deploy service-name --from-package` and I use a path that is nothing related to the `service-name` ( for example, the service in the template `foo` is a python app and I pass the path to a node app ).  Would that be an error?  or should azd just follow the order and deploy? (as per customer's request)?

**Review by vhvb1989 — APPROVED** [review]
LGTM, but please wait for @ellismg / @savannahostrowski approval.
I think this feature opens a wide range of possibilities that we should anticipate and document what customer should expect.

File: cli/azd/cmd/deploy.go
Comment (wbreza) [reply cli/azd/cmd/deploy.go:72]: I can update the text to be more generic like `package path`.  I can also update the help to dictate that it accepts whatever the output of the `package` command returns.

File: cli/azd/cmd/deploy.go
Comment (wbreza) [reply cli/azd/cmd/deploy.go:247]: Azd will not validate that the package was generated by azd or that it is the correct service / language.  It is up to the user to ensure they are passing a compatible package.

This supports scenarios where devs can build once, deploy multiple times promoting application package builds up through dev, test, prod environments.

The service target is responsible for ensuring the `--from-package` parameter is of the expected type.  Ex) app service will fail if not provided a valid zip file path.

Container apps will fail if the value is not a valid local docker image tag.

**Pull request merged**
